# Advanced Feature Engineering: Aggregating and Analyzing Grocery Category Amounts with SQL, Python, and PySpark

**Introduction:**

In this comprehensive data science project, our main focus is to showcase the process of advanced feature engineering using a combination of PySpark, Python, and SQL. We aim to demonstrate how to extract valuable insights and create multiple relevant variables from a dataset representing grocery category amounts in a simple and approachable manner.


**Data Description:**

The dataset used in this project represents the shopping data of three clients over a three-month period. It provides detailed information on the amount of money spent by each client on various categories of grocery items. Each entry includes the client's ID, the period of purchase, and the corresponding amounts spent on categories such as bakery, produce, dairy, meat, frozen foods, canned goods, beverages, snacks, condiments, and grains. It is important to note that this dataset is a mock dataset created solely for the purpose of demonstrating feature engineering techniques and does not reflect actual supermarket data.


**Project Goals:**

The main objective of this project is to illustrate how feature engineering can create new variables that capture important information and patterns within the dataset. By leveraging PySpark, Python, and SQL, we aim to showcase the process of manipulating and transforming the data to generate these new variables. These variables have the potential to provide deeper insights into customer behavior and enhance the accuracy of predictive models.


**Data Science Techniques:**

To achieve our project goals, we will leverage the capabilities of PySpark, Python, and SQL. These powerful tools allow us to perform various feature engineering techniques, such as aggregations, calculations, and data manipulation. By combining these techniques, we can create new variables that capture relevant information from the original features of the dataset. This process will enable us to uncover hidden patterns, understand customer preferences, and improve the overall analysis.


**Conclusions:** 

Please refer to the end of the project for detailed conclusions.

# Creating the Grocery Supermarket Dataset

In this data science portfolio project, we start by importing the necessary libraries and initializing Apache Spark using the "findspark" package. This ensures that Spark is ready to be used for our data processing tasks.

In [1]:
import findspark
findspark.init()
findspark.find()

'C:\\Users\\u632915\\.conda\\envs\\my-env\\lib\\site-packages\\pyspark'

First, we define the schema for our dataset, specifying the column names and data types. It's important to note that the dataset used in this code is a mock dataset created solely for the purpose of demonstrating feature engineering techniques and does not reflect actual supermarket data.

Then, we create a DataFrame named "grocery_category_amount_db" using the provided sample data and the defined schema. This DataFrame represents the grocery category amounts.

Additionally, we create a temporary view of the DataFrame, allowing us to easily perform SQL-like queries on the data.

Finally, we convert the PySpark DataFrame to a Python DataFrame using the toPandas() method. This conversion enables us to work with the data using familiar Python libraries and syntax, providing more flexibility and ease of analysis.

In [2]:
from pyspark.sql import SparkSession
from pyspark.sql.types import StructType, StructField, StringType, IntegerType

# Define the schema
schema = StructType([
    StructField('CLIENT_ID', StringType(), True),
    StructField('Period', StringType(), True),
    StructField('Bakery_amount', IntegerType(), True),
    StructField('Produce_amount', IntegerType(), True),
    StructField('Dairy_amount', IntegerType(), True),
    StructField('Meat_amount', IntegerType(), True),
    StructField('Frozen_foods_amount', IntegerType(), True),
    StructField('Canned_goods_amount', IntegerType(), True),
    StructField('Beverages_amount', IntegerType(), True),
    StructField('Snacks_amount', IntegerType(), True),
    StructField('Condiments_amount', IntegerType(), True),
    StructField('Grains_amount', IntegerType(), True)
])

# Create a sample DataFrame with mock data
data = [
    ('client_1', 'Jan-2023', 150, 200, 0, 80, 250, 220, 180, 300, 90, 120),
    ('client_1', 'Feb-2023', 0, 150, 150, 200, 150, 150, 150, 150, 200, 100),
    ('client_1', 'Mar-2023', 0, 120, 80, 150, 80, 120, 150, 250, 150, 150),
    ('client_2', 'Jan-2023', 1300, 10, 0, 0, 90, 0, 0, 0, 0, 0),
    ('client_2', 'Feb-2023', 1000, 30, 80, 0, 0, 0, 80, 0, 80, 0),
    ('client_2', 'Mar-2023', 1200, 40, 20, 0, 0, 0, 0, 0, 0, 0),
    ('client_3', 'Jan-2023', 200, 150, 150, 200, 150, 150, 150, 150, 200, 100),
    ('client_3', 'Feb-2023', 120, 120, 80, 150, 80, 0, 150, 250, 150, 150),
    ('client_3', 'Mar-2023', 0, 220, 180, 120, 90, 160, 120, 80, 200, 140)
]

# Create the SparkSession
spark = SparkSession.builder.appName("MyApp").getOrCreate()

# Create a database for grocery category amounts
grocery_category_amount_db = spark.createDataFrame(data, schema)

# Create a temporary view
grocery_category_amount_db.createOrReplaceTempView("grocery_category_amount")

# Show the database
# grocery_category_amount_db.show(1, vertical=True)

# Convert PySpark DataFrame to Python DataFrame
python_df = grocery_category_amount_db.toPandas()
python_df

CLIENT_ID    Period  Bakery_amount  Produce_amount  Dairy_amount  \
0  client_1  Jan-2023            150             200             0   
1  client_1  Feb-2023              0             150           150   
2  client_1  Mar-2023              0             120            80   
3  client_2  Jan-2023           1300              10             0   
4  client_2  Feb-2023           1000              30            80   
5  client_2  Mar-2023           1200              40            20   
6  client_3  Jan-2023            200             150           150   
7  client_3  Feb-2023            120             120            80   
8  client_3  Mar-2023              0             220           180   

   Meat_amount  Frozen_foods_amount  Canned_goods_amount  Beverages_amount  \
0           80                  250                  220               180   
1          200                  150                  150               150   
2          150                   80                  120               150   
3            0                   90                    0                 0   
4            0                    0                    0                80   
5            0                    0                    0                 0   
6          200                  150                  150               150   
7          150                   80                    0               150   
8          120                   90                  160               120   

   Snacks_amount  Condiments_amount  Grains_amount  
0            300                 90            120  
1            150                200            100  
2            250                150            150  
3              0                  0              0  
4              0                 80              0  
5              0                  0              0  
6            150                200            100  
7            250                150            150  
8             80                200            140

As we can see above, the objective is to analyze the expenditure of each customer on different categories of grocery items. Each entry contains the customer's ID, the purchase period, and the corresponding expenses for categories including bakery, produce, dairy, meat, frozen foods, canned goods, beverages, snacks, condiments, and grains. It is essential to note that this dataset is small and fictitious, created exclusively for showcasing feature engineering techniques. It does not represent real supermarket data.
Below, we will explore the process of aggregating grocery category amounts using PySpark for feature engineering.

# Feature Engineering: Aggregating Grocery Category Amounts with PySpark

First, we define the `calculate_total` function. This function accepts two parameters: `pVar` and `pName`. When calling the function as `calculate_total(pVar, pName)`, `pVar` represents a list of variables from the dataframe to which the function will be applied, while `pName` is the name assigned to the calculated field.

The purpose of this function is to generate SQL code that calculates the sum of all fields in the dataset that end with the suffix "_amount". To achieve this, the function iterates through the elements in `pVar` and constructs a partial SQL statement using the `coalesce` function. This function handles any null values by replacing them with zero.

The resulting SQL code is stored as a string in the `sql_total_amount` object. It includes the summed values of the relevant fields and assigns it an alias of "total_" followed by `pName`.

In the code snippet provided, we first retrieve the first row of the dataset `grocery_category_amount_df` using the `spark.sql` function. This dataset likely contains information about grocery category amounts.

Then, we call the `calculate_total` function with a list comprehension that extracts the column names ending with "_amount" from `grocery_category_amount_df.columns`, along with the category name 'amount'. The result is assigned to the `sql_total_amount` variable, which now holds the generated SQL code for calculating the total amount.

Overall, this code demonstrates the process of generating SQL code for aggregating grocery category amounts, specifically by summing the corresponding fields and handling null values using `coalesce`.

In [3]:
from pyspark.sql.functions import coalesce

# Define the function
def calculate_total(pVar, pName):
    sql = ","
    for h in pVar:
        if sql == ",":
            sql += "SUM(coalesce("+h+",0)"
        else: 
            sql += "+coalesce("+h+",0)"
    sql += ") as total_" + pName

    return sql

grocery_category_amount_df  = spark.sql("SELECT * FROM grocery_category_amount limit 1")
sql_total_amount = calculate_total([ x for x in grocery_category_amount_df.columns if x.endswith('_amount')], 'amount')
sql_total_amount

',SUM(coalesce(Bakery_amount,0)+coalesce(Produce_amount,0)+coalesce(Dairy_amount,0)+coalesce(Meat_amount,0)+coalesce(Frozen_foods_amount,0)+coalesce(Canned_goods_amount,0)+coalesce(Beverages_amount,0)+coalesce(Snacks_amount,0)+coalesce(Condiments_amount,0)+coalesce(Grains_amount,0)) as total_amount'

We create a dataframe called `total_grocery_category_amount_df` by executing an SQL query. In the query, we utilize the previously defined `sql_total_amount` object and group the results by the `CLIENT_ID` field, which is declared as `key_field: key_field = 'CLIENT_ID'`.

To execute the query, we use the `spark.sql()` function and pass it a multi-line SQL string. The string includes the `key_field` and `sql_total_amount` to select and calculate the total amount for each client. The `grocery_category_amount` table is used as the data source.

After executing the query, we display the resulting dataframe using the `show()` method. Setting `vertical=True` allows for a vertical display of the dataframe, making it easier to view all the columns.

In [4]:
key_field= 'CLIENT_ID'

# Use the result in a SQL query
total_grocery_category_amount_df = spark.sql("""SELECT """ + key_field + sql_total_amount + """ FROM grocery_category_amount
group by """+ key_field)
total_grocery_category_amount_df.show(vertical=True)

-RECORD 0----------------
 CLIENT_ID    | client_1 
 total_amount | 4240     
-RECORD 1----------------
 CLIENT_ID    | client_2 
 total_amount | 3930     
-RECORD 2----------------
 CLIENT_ID    | client_3 
 total_amount | 4160     



As observed in the above output, client_1 has incurred a total expenditure of 4240 over the past 3 months. Similarly, client_2 has accumulated a total expense of 3930 during the same period, while client_3 has spent a total amount of 4160 over the last 3 months.

# Feature Engineering: Calculating Multiple Statistics for Grocery Category Amounts with PySpark

Similarly to the `calculate_total` function, we can create a `calculate_derivatives` function. This function takes one parameter, `pVar`, which represents a list of variables from the dataframe to which the function will be applied. The purpose of this function is to generate SQL code that calculates various statistics for each field in the dataset.

In the `calculate_derivatives` function, we iterate through the elements in `pVar` and construct a SQL string. This string includes calculations for the average (`AVG`), sum (`SUM`), minimum (`MIN`), maximum (`MAX`), and count of periods where the value is greater than zero for each field. The `coalesce` function handles any null values by replacing them with zero.

The resulting SQL code is returned as a string.

In the provided code snippet, we first retrieve the first row of the `grocery_category_amount_df` dataset using the `spark.sql` function. This dataset likely contains information about grocery category amounts.

Then, we call the `calculate_derivatives` function with a list comprehension that extracts the column names ending with "_amount" from `grocery_category_amount_df.columns`. The result is assigned to the `sql_derivatives` variable, which now holds the generated SQL code for calculating the desired statistics.

Overall, this code showcases the creation of a function for generating SQL code to calculate various statistics, such as average, sum, minimum, maximum, and the count of periods with non-zero values for each field in the dataset.

In [5]:
def calculate_derivatives(pVar):
    sql = ""
    for h in pVar:
        sql += ",AVG(coalesce("+h+",0)) as " + h + "_avg " + \
               ",SUM(coalesce("+h+",0)) as " + h + "_sum " + \
               ",min(coalesce("+h+",0)) as " + h + "_min " + \
               ",max(coalesce("+h+",0)) as " + h + "_max " + \
               " ,count(distinct case when coalesce("+h+",0) > 0 then Period end  ) as " + h + "_periods "        
    return sql

grocery_category_amount_df = spark.sql("SELECT * FROM grocery_category_amount limit 1")
sql_derivatives = calculate_derivatives([ x for x in grocery_category_amount_df.columns if x.endswith('_amount')])
sql_derivatives

',AVG(coalesce(Bakery_amount,0)) as Bakery_amount_avg ,SUM(coalesce(Bakery_amount,0)) as Bakery_amount_sum ,min(coalesce(Bakery_amount,0)) as Bakery_amount_min ,max(coalesce(Bakery_amount,0)) as Bakery_amount_max  ,count(distinct case when coalesce(Bakery_amount,0) > 0 then Period end  ) as Bakery_amount_periods ,AVG(coalesce(Produce_amount,0)) as Produce_amount_avg ,SUM(coalesce(Produce_amount,0)) as Produce_amount_sum ,min(coalesce(Produce_amount,0)) as Produce_amount_min ,max(coalesce(Produce_amount,0)) as Produce_amount_max  ,count(distinct case when coalesce(Produce_amount,0) > 0 then Period end  ) as Produce_amount_periods ,AVG(coalesce(Dairy_amount,0)) as Dairy_amount_avg ,SUM(coalesce(Dairy_amount,0)) as Dairy_amount_sum ,min(coalesce(Dairy_amount,0)) as Dairy_amount_min ,max(coalesce(Dairy_amount,0)) as Dairy_amount_max  ,count(distinct case when coalesce(Dairy_amount,0) > 0 then Period end  ) as Dairy_amount_periods ,AVG(coalesce(Meat_amount,0)) as Meat_amount_avg ,SUM(coales

We generate a DataFrame called `grouped_grocery_category_amount` by executing an SQL query. In this query, we utilize the previously defined `sql_total_amount` and `sql_derivatives` objects. The results are grouped by the `CLIENT_ID` field, which we declare as `key_field: key_field = 'CLIENT_ID'`.

To execute the query, we use the `spark.sql()` function and provide it with a multi-line SQL string. This string includes the `key_field`, `sql_total_amount`, and `sql_derivatives` to select and calculate the total amount and various statistics for each client. The `grocery_category_amount` table serves as the data source for the query.

After executing the query, we convert the resulting DataFrame to a Python DataFrame using the `toPandas()` method. This allows us to work with the data in a familiar Python environment.

To display all columns of the `python_df2` DataFrame, we use the `pd.set_option()` function from the Pandas library to set the display options for maximum columns. By setting `display.max_columns` to `None`, all columns in the DataFrame will be shown when we print it.

Overall, this code demonstrates the utilization of SQL queries in PySpark to group the grocery category amounts by client and calculate various statistics. It further showcases the seamless integration between PySpark and Pandas, allowing us to work with the results in a Python-friendly manner.

In [6]:
import pandas as pd

key_field= 'CLIENT_ID'

# Use the result in a SQL query
grouped_grocery_category_amount = spark.sql("""SELECT """ + key_field + sql_derivatives + sql_total_amount + """ FROM grocery_category_amount
group by """+ key_field)

# grouped_grocery_category_amount.show(1,vertical=True)

# Convert PySpark DataFrame to Python DataFrame
python_df2 = grouped_grocery_category_amount.toPandas()
pd.set_option('display.max_columns', None)  # Set maximum columns to display
python_df2

CLIENT_ID  Bakery_amount_avg  Bakery_amount_sum  Bakery_amount_min  \
0  client_3         106.666667                320                  0   
1  client_1          50.000000                150                  0   
2  client_2        1166.666667               3500               1000   

   Bakery_amount_max  Bakery_amount_periods  Produce_amount_avg  \
0                200                      2          163.333333   
1                150                      1          156.666667   
2               1300                      3           26.666667   

   Produce_amount_sum  Produce_amount_min  Produce_amount_max  \
0                 490                 120                 220   
1                 470                 120                 200   
2                  80                  10                  40   

   Produce_amount_periods  Dairy_amount_avg  Dairy_amount_sum  \
0                       3        136.666667               410   
1                       3         76.666667               230   
2                       3         33.333333               100   

   Dairy_amount_min  Dairy_amount_max  Dairy_amount_periods  Meat_amount_avg  \
0                80               180                     3       156.666667   
1                 0               150                     2       143.333333   
2                 0                80                     2         0.000000   

   Meat_amount_sum  Meat_amount_min  Meat_amount_max  Meat_amount_periods  \
0              470              120              200                    3   
1              430               80              200                    3   
2                0                0                0                    0   

   Frozen_foods_amount_avg  Frozen_foods_amount_sum  Frozen_foods_amount_min  \
0               106.666667                      320                       80   
1               160.000000                      480                       80   
2                30.000000                       90                        0   

   Frozen_foods_amount_max  Frozen_foods_amount_periods  \
0                      150                            3   
1                      250                            3   
2                       90                            1   

   Canned_goods_amount_avg  Canned_goods_amount_sum  Canned_goods_amount_min  \
0               103.333333                      310                        0   
1               163.333333                      490                      120   
2                 0.000000                        0                        0   

   Canned_goods_amount_max  Canned_goods_amount_periods  Beverages_amount_avg  \
0                      160                            2            140.000000   
1                      220                            3            160.000000   
2                        0                            0             26.666667   

   Beverages_amount_sum  Beverages_amount_min  Beverages_amount_max  \
0                   420                   120                   150   
1                   480                   150                   180   
2                    80                     0                    80   

   Beverages_amount_periods  Snacks_amount_avg  Snacks_amount_sum  \
0                         3         160.000000                480   
1                         3         233.333333                700   
2                         1           0.000000                  0   

   Snacks_amount_min  Snacks_amount_max  Snacks_amount_periods  \
0                 80                250                      3   
1                150                300                      3   
2                  0                  0                      0   

   Condiments_amount_avg  Condiments_amount_sum  Condiments_amount_min  \
0             183.333333                    550                    150   
1             146.666667                    440                     90   
2              26.666667               

In the provided output, we can see that client_3 has an average spending of 106.66 in the Bakery category over the last 3 periods. The total amount spent in Bakery over the 3 periods was 320. The minimum amount spent in Bakery was 0, and the maximum amount was 200. Client_3 made purchases in the Bakery category in 2 out of the 3 periods.

Similarly, client_1 has an average spending of 50 in the Bakery category over the last 3 periods. The total amount spent in Bakery over the 3 periods was 150. The minimum amount spent in Bakery was 0, and the maximum amount was 150. Client_1 made purchases in the Bakery category in 1 out of the 3 periods.

Additionally, client_2 has an average spending of 1166.66 in the Bakery category over the last 3 periods. The total amount spent in Bakery over the 3 periods was 3500. The minimum amount spent in Bakery was 1000, and the maximum amount was $1300. Client_2 made purchases in the Bakery category in all 3 periods.

These insights provide a summary of the spending behavior for each client in the Bakery category, including average amounts, total spending, minimum and maximum amounts, and the number of periods in which purchases were made. Similar calculations are performed for other categories such as produce, dairy, meat, frozen foods, canned goods, beverages, snacks, condiments, and grains.

# Feature Engineering: Calculating the Percentages of the Total Amount for Grocery Category Amounts with PySpark

Before utilizing the `grouped_grocery_category_amount` dataframe in SQL queries, we establish a temporary view named "grouped_grocery_category_amount" through the `createOrReplaceTempView` method. This action enables us to reference the dataframe by its designated view name in subsequent SQL queries, simplifying the syntax and enhancing readability.

Like the `calculate_total` and `calculate_derivatives` functions, we define the `calculate_percentages` function. This function takes two parameters: `pVar` and `pName`. When calling the function as `calculate_percentages(pVar, pName)`, `pVar` represents a list of variables from the dataframe to which the function will be applied, while `pName` is the name of the total amount field created previously with `calculate_total`.

The purpose of this function is to generate SQL code that calculates the percentage of the total amount for each category of all fields in the dataset that end with the suffix "_sum". To achieve this, the function iterates through the elements in `pVar` and constructs a partial SQL statement.

The resulting SQL code is stored as a string in the `sql_percent_amount` object.

In the provided code snippet, we first retrieve the first row of the `grouped_grocery_category_amount_df` dataset using the `spark.sql` function. This dataset likely contains information about grocery category amounts.

Then, we call the `calculate_percentages` function with a list comprehension that extracts the column names ending with "_sum" from `grouped_grocery_category_amount_df.columns`. The result is assigned to the `sql_percent_amount` variable, which now holds the generated SQL code for calculating the desired percentages.

Overall, this code demonstrates the process of generating SQL code to calculate the percentages of the total amount for each category in the dataset based on the summed values.

In [7]:
# Create a temporary view
grouped_grocery_category_amount.createOrReplaceTempView("grouped_grocery_category_amount")

In [8]:
# % of TOTAL_AMOUNT

def calculate_percentages(pVar, pName):
    sql2 = ""
    for h in pVar:
        sql2 += ", round("+h+" / case when total_" + pName + """ = 0 then 1 else total_""" + pName + " end, 3) as " + h + "_perc"
    return sql2

grouped_grocery_category_amount_df = spark.sql("SELECT * FROM grouped_grocery_category_amount limit 1")
sql_percent_amount = calculate_percentages([ x for x in grouped_grocery_category_amount_df.columns if (x.endswith('_sum'))], 'amount')
sql_percent_amount

', round(Bakery_amount_sum / case when total_amount = 0 then 1 else total_amount end, 3) as Bakery_amount_sum_perc, round(Produce_amount_sum / case when total_amount = 0 then 1 else total_amount end, 3) as Produce_amount_sum_perc, round(Dairy_amount_sum / case when total_amount = 0 then 1 else total_amount end, 3) as Dairy_amount_sum_perc, round(Meat_amount_sum / case when total_amount = 0 then 1 else total_amount end, 3) as Meat_amount_sum_perc, round(Frozen_foods_amount_sum / case when total_amount = 0 then 1 else total_amount end, 3) as Frozen_foods_amount_sum_perc, round(Canned_goods_amount_sum / case when total_amount = 0 then 1 else total_amount end, 3) as Canned_goods_amount_sum_perc, round(Beverages_amount_sum / case when total_amount = 0 then 1 else total_amount end, 3) as Beverages_amount_sum_perc, round(Snacks_amount_sum / case when total_amount = 0 then 1 else total_amount end, 3) as Snacks_amount_sum_perc, round(Condiments_amount_sum / case when total_amount = 0 then 1 els

We create a DataFrame named `grouped_grocery_category_amount_V2` by executing an SQL query. In this query, we utilize the previously defined `sql_percent_amount` object. The results are grouped by the `CLIENT_ID` field, which we declare as `key_field: key_field = 'CLIENT_ID'`.

To execute the query, we use the `spark.sql()` function and provide it with a multi-line SQL string. This string includes the `key_field` and `sql_percent_amount`. The `grouped_grocery_category_amount` DataFrame serves as the data source for the query.

After executing the query, we convert the resulting DataFrame to a Python DataFrame using the `toPandas()` method. This conversion allows us to work with the data in a familiar Python environment and utilize various Python libraries and functions for further analysis and visualization.

Overall, this code demonstrates the usage of SQL queries in PySpark to generate a new DataFrame, `grouped_grocery_category_amount_V2`, by calculating the percentages of the total amount for each category. The resulting DataFrame is then converted to a Python DataFrame for easier data manipulation and exploration.

In [9]:
# Use the result in a SQL query
grouped_grocery_category_amount_V2 = spark.sql("""SELECT """ + key_field + sql_percent_amount + """ FROM grouped_grocery_category_amount """)

# Convert PySpark DataFrame to Python DataFrame
python_df3 = grouped_grocery_category_amount_V2.toPandas()
# pd.set_option('display.max_columns', None)  # Set maximum columns to display
python_df3

CLIENT_ID  Bakery_amount_sum_perc  Produce_amount_sum_perc  \
0  client_1                   0.035                    0.111   
1  client_2                   0.891                    0.020   
2  client_3                   0.077                    0.118   

   Dairy_amount_sum_perc  Meat_amount_sum_perc  Frozen_foods_amount_sum_perc  \
0                  0.054                 0.101                         0.113   
1                  0.025                 0.000                         0.023   
2                  0.099                 0.113                         0.077   

   Canned_goods_amount_sum_perc  Beverages_amount_sum_perc  \
0                         0.116                      0.113   
1                         0.000                      0.020   
2                         0.075                      0.101   

   Snacks_amount_sum_perc  Condiments_amount_sum_perc  Grains_amount_sum_perc  
0                   0.165                       0.104                   0.087  
1                   0.000                       0.020                   0.000  
2                   0.115                       0.132                   0.094

In the provided output, we can observe the spending percentages for client_1, client_2, and client_3 in different categories over the last 3 months.

For client_1, we see that they spent approximately 3.5% of their total expenditure in the Bakery category, 1.11% in the Produce category, and allocated portions of their spending to other categories as well. The percentages reflect the proportion of client_1's total expenditure allocated to each category.

Similarly, for client_2 and client_3, we can analyze their spending patterns across the different categories. The same logic applies, where we observe the percentages of total spending allocated to each category for client_2 and client_3.

These percentages provide valuable insights into the distribution of spending for each client across various categories. It allows us to understand their preferences and patterns of expenditure. Importantly, the sum of these percentages for each client adds up to 100%, indicating that all of their spending has been accounted for across the analyzed categories.

# Feature Engineering: Calculating the Flag for High Spending Customers in Specific Supermarket Categories (PySpark)

Before using the `grouped_grocery_category_amount_V2` DataFrame in SQL queries, we create a temporary view named "grouped_grocery_category_amount_V2" using the `createOrReplaceTempView` method. This step allows us to refer to the DataFrame by its assigned view name in subsequent SQL queries, simplifying the syntax and improving readability.

Similar to the `calculate_total`, `calculate_derivatives`, and `calculate_percentages` functions, we also define the `MoreThan80pInCategory` function. This function accepts two parameters: `pVar` and `pName`. When calling the function as `MoreThan80pInCategory(pVar, pName)`, `pVar` represents a list of variables from the DataFrame to which the function will be applied, while `pName` denotes the calculated field name.

The purpose of this function is to generate SQL code that calculates the flag (0 or 1) indicating whether a customer spends more than 80% of the total amount in at least one specific supermarket category. This helps identify customers who consistently spend in a particular category or across multiple categories. To accomplish this, the function iterates through the elements in `pVar` and constructs a partial SQL statement.

The resulting SQL code is stored as a string in the `sql_MoreThan80pInCategory` object.

In the provided code snippet, we first retrieve the first row of the `grouped_grocery_category_amount_V2` dataset using the `spark.sql` function.

Next, we invoke the `MoreThan80pInCategory` function with a list comprehension that extracts the column names ending with "_perc" from `grouped_grocery_category_amount_V2.columns`. The generated SQL code for calculating the flag is assigned to the `sql_MoreThan80pInCategory` variable.

Overall, this code demonstrates the process of generating SQL code to calculate the flag variable for each category in the dataset based on customers' spending patterns, particularly whether they exceed 80% of the total amount in at least one category.

In [10]:
# Create a temporary view
grouped_grocery_category_amount_V2.createOrReplaceTempView("grouped_grocery_category_amount_V2")

In [11]:
# Flag Customers consuming more than 80% in at least one category

def MoreThan80pInCategory(pVar, pName):
    sql = ", CASE WHEN "
    for p in pVar:
        pp = " case when " + p + " > 0.8 then 1 else 0 end"
        if sql == ", CASE WHEN ":
            sql += pp
        else:
            sql += " + " + pp
    sql += ">0 THEN 1 ELSE 0 END as FLAG_MoreThan80p_" + pName
    return sql

grouped_grocery_category_amount_V2_df = spark.sql("SELECT * FROM grouped_grocery_category_amount_V2 limit 1")
sql_MoreThan80pInCategory = MoreThan80pInCategory([ x for x in grouped_grocery_category_amount_V2.columns if (x.endswith('_perc'))], 'amount')
sql_MoreThan80pInCategory

', CASE WHEN  case when Bakery_amount_sum_perc > 0.8 then 1 else 0 end +  case when Produce_amount_sum_perc > 0.8 then 1 else 0 end +  case when Dairy_amount_sum_perc > 0.8 then 1 else 0 end +  case when Meat_amount_sum_perc > 0.8 then 1 else 0 end +  case when Frozen_foods_amount_sum_perc > 0.8 then 1 else 0 end +  case when Canned_goods_amount_sum_perc > 0.8 then 1 else 0 end +  case when Beverages_amount_sum_perc > 0.8 then 1 else 0 end +  case when Snacks_amount_sum_perc > 0.8 then 1 else 0 end +  case when Condiments_amount_sum_perc > 0.8 then 1 else 0 end +  case when Grains_amount_sum_perc > 0.8 then 1 else 0 end>0 THEN 1 ELSE 0 END as FLAG_MoreThan80p_amount'

We generate a DataFrame called `grouped_grocery_category_amount_V3` by executing an SQL query. This query utilizes the previously defined `sql_MoreThan80pInCategory` object. The query groups the data by the `CLIENT_ID` field, which we declare as `key_field: key_field = 'CLIENT_ID'`.

To execute the query, we use the `spark.sql()` function and provide it with a multi-line SQL string. This string includes the `key_field` and `sql_MoreThan80pInCategory`. The `grouped_grocery_category_amount_V2` DataFrame serves as the data source for the query.

After executing the query, we convert the resulting DataFrame to a Python DataFrame using the `toPandas()` method. This conversion allows us to work with the data in a familiar Python environment and leverage various Python libraries and functions for further analysis and visualization.

Overall, this code demonstrates the utilization of SQL queries in PySpark to generate a new DataFrame, `grouped_grocery_category_amount_V3`, by calculating the flag (0 or 1) indicating whether a customer spends more than 80% of the total amount in at least one specific supermarket category. The resulting DataFrame is then converted to a Python DataFrame for easier data manipulation and exploration.

In [12]:
# Use the result in a SQL query
grouped_grocery_category_amount_V3 = spark.sql("""SELECT """ + key_field + sql_MoreThan80pInCategory + """,a.* FROM grouped_grocery_category_amount_V2 a """)
# Convert PySpark DataFrame to Python DataFrame
python_df4 = grouped_grocery_category_amount_V3.toPandas()
# pd.set_option('display.max_columns', None)  # Set maximum columns to display
python_df4

CLIENT_ID  FLAG_MoreThan80p_amount CLIENT_ID  Bakery_amount_sum_perc  \
0  client_1                        0  client_1                   0.035   
1  client_2                        1  client_2                   0.891   
2  client_3                        0  client_3                   0.077   

   Produce_amount_sum_perc  Dairy_amount_sum_perc  Meat_amount_sum_perc  \
0                    0.111                  0.054                 0.101   
1                    0.020                  0.025                 0.000   
2                    0.118                  0.099                 0.113   

   Frozen_foods_amount_sum_perc  Canned_goods_amount_sum_perc  \
0                         0.113                         0.116   
1                         0.023                         0.000   
2                         0.077                         0.075   

   Beverages_amount_sum_perc  Snacks_amount_sum_perc  \
0                      0.113                   0.165   
1                      0.020                   0.000   
2                      0.101                   0.115   

   Condiments_amount_sum_perc  Grains_amount_sum_perc  
0                       0.104                   0.087  
1                       0.020                   0.000  
2                       0.132                   0.094

As we can see in the table above, client 2 has a FLAG_MoreThan80p_amount = 1 because this client spends 89% of their total expenditure in the Bakery category over the last 3 periods or months.

# Feature Engineering: Generating Ratio Variables for Grocery Category Amounts with PySpark

Finally, we can generate ratio variables. Similar to the `calculate_total`, `calculate_derivatives`, `calculate_percentages`, and `MoreThan80pInCategory` functions, we also define the `CalculateRatioComparison` function. This function accepts one parameter: `pVar`. When calling the function as `CalculateRatioComparison(pVar)`, `pVar` represents a list of variables from the DataFrame to which the function will be applied.

The purpose of this function is to generate SQL code that calculates the ratio between supermarket categories. This helps in creating interaction variables. To achieve this, the function iterates through the elements in `pVar` and constructs a partial SQL statement.

The resulting SQL code is stored as a string in the `sql_ratio_amount` object.

In the provided code snippet, we first retrieve the first row of the `grouped_grocery_category_amount` dataset using the `spark.sql` function.

Next, we invoke the `CalculateRatioComparison` function with a list comprehension that extracts the column names ending with "_sum" from `grouped_grocery_category_amount.columns`. This generates the SQL code for calculating the ratio between supermarket categories.

Overall, this code demonstrates the process of generating SQL code to calculate the ratio variables for each category in the dataset based on their summed values.

In [13]:
def CalculateRatioComparison(pVar):
    sql = ""
    for h in pVar:
          for h2 in pVar:
                sql += ", round("+h+" / case when " + h2 + " = 0 then 1 else " + h2 + " end, 3) as " + h + "_vs_" + h2
    return  sql

grouped_grocery_category_amount_df = spark.sql("SELECT * FROM grouped_grocery_category_amount limit 1")
sql_ratio_amount = CalculateRatioComparison([ x for x in grouped_grocery_category_amount.columns if (x.endswith('_sum'))])
sql_ratio_amount

', round(Bakery_amount_sum / case when Bakery_amount_sum = 0 then 1 else Bakery_amount_sum end, 3) as Bakery_amount_sum_vs_Bakery_amount_sum, round(Bakery_amount_sum / case when Produce_amount_sum = 0 then 1 else Produce_amount_sum end, 3) as Bakery_amount_sum_vs_Produce_amount_sum, round(Bakery_amount_sum / case when Dairy_amount_sum = 0 then 1 else Dairy_amount_sum end, 3) as Bakery_amount_sum_vs_Dairy_amount_sum, round(Bakery_amount_sum / case when Meat_amount_sum = 0 then 1 else Meat_amount_sum end, 3) as Bakery_amount_sum_vs_Meat_amount_sum, round(Bakery_amount_sum / case when Frozen_foods_amount_sum = 0 then 1 else Frozen_foods_amount_sum end, 3) as Bakery_amount_sum_vs_Frozen_foods_amount_sum, round(Bakery_amount_sum / case when Canned_goods_amount_sum = 0 then 1 else Canned_goods_amount_sum end, 3) as Bakery_amount_sum_vs_Canned_goods_amount_sum, round(Bakery_amount_sum / case when Beverages_amount_sum = 0 then 1 else Beverages_amount_sum end, 3) as Bakery_amount_sum_vs_Beverag

We generate a DataFrame called `grouped_grocery_category_amount_V4` by executing an SQL query. This query utilizes the previously defined `sql_ratio_amount` object. The query groups the data by the `CLIENT_ID` field, which we declare as `key_field`: `key_field = 'CLIENT_ID'`.

To execute the query, we use the `spark.sql()` function and provide it with a multi-line SQL string. This string includes the `key_field` and `sql_ratio_amount`. The `grouped_grocery_category_amount` DataFrame serves as the data source for the query.

After executing the query, we display the first row of the resulting DataFrame using the `show()` function. The `vertical=True` parameter allows us to display the row vertically, providing a clear view of the calculated ratio variables.

Overall, this code demonstrates the usage of SQL queries in PySpark to generate a new DataFrame, `grouped_grocery_category_amount_V4`, by calculating the ratio between supermarket categories. The resulting DataFrame is then displayed to examine the calculated ratio variables.

In [14]:
# Use the result in a SQL query
grouped_grocery_category_amount_V4 = spark.sql("""SELECT """ + key_field + sql_ratio_amount + """ FROM grouped_grocery_category_amount """)
grouped_grocery_category_amount_V4.show(1,vertical=True)

-RECORD 0------------------------------------------------------
 CLIENT_ID                                          | client_1 
 Bakery_amount_sum_vs_Bakery_amount_sum             | 1.0      
 Bakery_amount_sum_vs_Produce_amount_sum            | 0.319    
 Bakery_amount_sum_vs_Dairy_amount_sum              | 0.652    
 Bakery_amount_sum_vs_Meat_amount_sum               | 0.349    
 Bakery_amount_sum_vs_Frozen_foods_amount_sum       | 0.313    
 Bakery_amount_sum_vs_Canned_goods_amount_sum       | 0.306    
 Bakery_amount_sum_vs_Beverages_amount_sum          | 0.313    
 Bakery_amount_sum_vs_Snacks_amount_sum             | 0.214    
 Bakery_amount_sum_vs_Condiments_amount_sum         | 0.341    
 Bakery_amount_sum_vs_Grains_amount_sum             | 0.405    
 Produce_amount_sum_vs_Bakery_amount_sum            | 3.133    
 Produce_amount_sum_vs_Produce_amount_sum           | 1.0      
 Produce_amount_sum_vs_Dairy_amount_sum             | 2.043    
 Produce_amount_sum_vs_Meat_amount_sum  

In the output above, we can see the ratios calculated for client_1 over the last 3 months. For client_1, the ratio between Bakery and Produce is 0.319, indicating that the amount spent on Bakery is approximately 31.9% of the amount spent on Produce during this period.

Similarly, the ratio between Bakery and Dairy is 0.652, implying that the amount spent on Bakery is about 65.2% of the amount spent on Dairy during this period.

Lastly, the ratio between Bakery and Meat is 0.349, suggesting that the amount spent on Bakery is approximately 34.9% of the amount spent on Meat during this period.

These ratios provide insights into the relative distribution of spending between Bakery and other categories for client_1. The same ratios have also been calculated for client_2 and client_3, allowing us to analyze their spending patterns and compare them across the same categories.

# Feature Engineering: Final Dataset with All Created Variables

In this section, we performed the feature engineering process to create the final dataset called "grouped_grocery_category_amount_final." This dataset includes all the variables we generated throughout the project using data manipulation techniques in PySpark and SQL.

The code used to create the final dataset is as follows:

In [15]:
grouped_grocery_category_amount_final = spark.sql("""SELECT grouped_grocery_category_amount.* """+ sql_MoreThan80pInCategory
+ """ FROM (SELECT A.* """+ sql_ratio_amount  + sql_percent_amount  + """ FROM (SELECT """ + key_field + 
sql_derivatives + sql_total_amount + """ FROM grocery_category_amount group by """+ 
key_field + """)A)grouped_grocery_category_amount """)

grouped_grocery_category_amount_final.show(1,vertical=True)

-RECORD 0----------------------------------------------------------------
 CLIENT_ID                                          | client_3           
 Bakery_amount_avg                                  | 106.66666666666667 
 Bakery_amount_sum                                  | 320                
 Bakery_amount_min                                  | 0                  
 Bakery_amount_max                                  | 200                
 Bakery_amount_periods                              | 2                  
 Produce_amount_avg                                 | 163.33333333333334 
 Produce_amount_sum                                 | 490                
 Produce_amount_min                                 | 120                
 Produce_amount_max                                 | 220                
 Produce_amount_periods                             | 3                  
 Dairy_amount_avg                                   | 136.66666666666666 
 Dairy_amount_sum                     

Above is a preview of the final dataset with one observation.
This final dataset contains all the variables generated through the feature engineering process, providing a comprehensive and enriched view of grocery category expenditure data. Its creation enables us to conduct in-depth analysis and build more accurate and effective Data Science models.

# Conclusions:

This data science project showcased the power of advanced feature engineering using PySpark, Python, and SQL. We explored the transformation of a dataset with 10 variables representing different grocery categories, including bakery, produce, dairy, meat, frozen foods, canned goods, beverages, snacks, condiments, and grains. The original table was structured at the CLIENT_ID and Period level, and through advanced feature engineering, we generated a final dataset at the CLIENT_ID level, aggregating data for the last 3 periods.

Throughout the project, we achieved the following:

- A) Created a new variable representing the "Total Amount expenditure over the past 3 months," aggregating the amounts spent across all grocery categories for each client. This provided a comprehensive overview of their overall spending during the specified period.

- B) Calculated various statistics, such as the average (AVG), sum (SUM), minimum (MIN), maximum (MAX), and count of periods where the value is greater than zero for each grocery category. These calculations provided insights into the average, total, and frequency of spending for each category over the last 3 periods.

- C) Computed the percentages of the Total Amount for each grocery category, enabling a comparison of the contribution of each category to the overall spending. This analysis shed light on the relative importance of different grocery categories for each client within the specified timeframe.

- D) Determined a flag for high spending customers in specific supermarket categories. By setting a threshold, we identified individuals who consistently spent above that threshold in particular categories, revealing their significant purchasing behavior.

- E) Generated ratio variables for the grocery categories, allowing for comparisons between categories. These ratios provided insights into the proportion of spending allocated to each grocery category and helped identify potential relationships or trends.

By leveraging advanced feature engineering techniques, we transformed the original dataset into a more informative and valuable resource for analysis at the CLIENT_ID level, considering the last 3 periods. The generated variables captured important patterns and behaviors related to grocery category spending, empowering data-driven decision-making and enhancing predictive modeling accuracy.

One notable quality of these derived variables is their stability over time. By considering data from the last 3 months, we capture more consistent trends and patterns in customer behavior, providing more robust features for our data science model.

Furthermore, these derived variables allow us to incorporate the temporal aspect of the data. By analyzing patterns and trends over the last 3 months, we can capture short-term changes in customer preferences and adjust our models accordingly.

In conclusion, this project highlights the importance of advanced feature engineering in data science. By utilizing PySpark, Python, and SQL, we demonstrated how to leverage these techniques to extract insights and create valuable variables from grocery category spending data. These techniques enable data scientists to uncover hidden patterns, understand customer preferences, and make informed business decisions based on a comprehensive analysis of their data, considering the CLIENT_ID level and aggregating data for the last 3 periods. The derived variables offer stability over time and incorporate the temporal aspect of the data, enhancing the accuracy and effectiveness of our data science models.